src/repositories/leaderboard_repository.rs

```diff
@@ -1,65 +1,108 @@
+
+#[derive(Clone)]
+pub struct LeaderboardRepository {
+    db: Arc<Session>,
+    insert_query: PreparedStatement,
+    find_by_partition: PreparedStatement,
+    delete_query: PreparedStatement,
+}
+
+impl LeaderboardRepository {
+    pub async fn new(db: Arc<Session>) -> Self {
+        let insert_query = db
+            .prepare(LEADERBOARD_INSERT_QUERY)
+            .await
+            .expect("[LEADERBOARD REPOSITORY] Error preparing insert query");
+        let mut find_by_partition = db
+            .prepare(LEADERBOARD_FIND_BY_PARTITION_KEYS_QUERY)
+            .await
+            .expect("[LEADERBOARD REPOSITORY] Error preparing find query");
+        find_by_partition.set_page_size(2);
+
+        let delete_query = db
+            .prepare(LEADERBOARD_DELETE_PREVIOUS_SCORE_QUERY)
+            .await
+            .expect("[LEADERBOARD REPOSITORY] Error preparing delete query");
+
+        LeaderboardRepository {
+            db,
+            insert_query,
+            find_by_partition,
+            delete_query,
+        }
+    }
+
+    pub async fn insert(&self, payload: SubmissionDTO) -> anyhow::Result<Leaderboard> {
+        let leaderboard = Leaderboard::from_request(payload);
+
+        self.db
+            .execute_unpaged(&self.insert_query, &leaderboard)
+            .await?;
+
+        Ok(leaderboard)
+    }
+
+    pub async fn delete_previous_score(&self, payload: &SubmissionDTO) -> anyhow::Result<()> {
+        self.db
+            .execute_unpaged(
+                &self.delete_query,
+                (
+                    &payload.song_id,
+                    &payload.player_id,
+                    &payload.difficulty,
+                    &payload.instrument,
+                    &payload.modifiers,
+                    &payload.score,
+                ),
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn get_leaderboard(
+        &self,
+        song_id: String,
+        filters: LeaderboardRequest,
+    ) -> anyhow::Result<(Vec<Leaderboard>, PagingStateResponse)> {
+        let paging_state = PagingState::start();
+
+        let (response, paging_state_response) = self
+            .db
+            .execute_single_page(
+                &self.find_by_partition,
+                (
+                    song_id,
+                    filters.modifiers,
+                    filters.difficulty,
+                    filters.instrument,
+                ),
+                paging_state,
+            )
+            .await?;
+
+        let response = response
+            .into_rows_result()
+            .expect("Failed to get rows result");
+        let rows = response.rows::<Leaderboard>().expect("Failed to get rows");
+        let mut result: Vec<Leaderboard> = Vec::new();
+
+        for row in rows {
+            result.push(row?);
+        }
+
+        Ok((result, paging_state_response))
```

Comment:
Nit: You should be able to collect into a vector instead of pushing manually. Something like this:
```rust
let result = response
    .rows::<Leaderboard>()
    .expect("Failed to get rows")
    .collect::<Result<Vec<Leaderboard>, DeserializationError>>()?;
```